components/cli/pkg/commands/image/test.go

```diff
@@ -254,196 +266,98 @@ func startTestCellInstance(imageDir string, instanceName string, runningNode *de
-	}
-	if containsTestFunction {
-		tempTestFileName, err := util.CreateTempExecutableBalFile(balFilePath, "test")
+		// Change working dir to Bal project
+		err = os.Chdir(testsRoot)
 		if err != nil {
-			util.ExitWithErrorMessage("Error executing ballerina file", err)
+			return fmt.Errorf("error occurred while changing working directory, %s", err)
 		}
-		// Preparing the run command arguments
-		cmdArgs := []string{"run"}
-		var startDependenciesFlag = "false"
-		if startDependencies {
-			startDependenciesFlag = "true"
-		}
-		var shareDependenciesFlag = "false"
-		if shareDependencies {
-			shareDependenciesFlag = "true"
-		}
-		fileCopyError := util.CopyFile(tempTestFileName, filepath.Join(balModule, filepath.Base(tempTestFileName)))
-		if fileCopyError != nil {
-			util.ExitWithErrorMessage("Error occurred while copying temp source bal file", err)
-		}
-		cmdArgs = append(cmdArgs, filepath.Base(balModule), "test", string(iName), string(dependencyLinksJson), startDependenciesFlag, shareDependenciesFlag)
-		util.CleanAndCreateDir(filepath.Join(currentDir, "target", "logs"))
-
-		defer os.Remove(imageDir)
-		cmd.Env = os.Environ()
 
-		if debug {
-			SetupDebugConfigs(string(iName), verboseMode, imageDir, string(dependencyLinksJson), envVars, assumeYes)
-		} else {
-			cmd.Env = append(cmd.Env, constants.CelleryImageDirEnvVar+"="+imageDir)
-			cmd.Env = append(cmd.Env, fmt.Sprintf("DEBUG_MODE=%s", verboseMode))
-			cmd.Env = append(cmd.Env, fmt.Sprintf("IMAGE_NAME=%s", string(iName)))
-			cmd.Env = append(cmd.Env, fmt.Sprintf("DEPENDENCY_LINKS=%s", string(dependencyLinksJson)))
-			cmd.Env = append(cmd.Env, fmt.Sprintf("TEST_MODULE=%s", filepath.Base(balModule)))
-			for _, envVar := range envVars {
-				cmd.Env = append(cmd.Env, fmt.Sprintf(envVar.Key+"=%s", envVar.Value))
+		isExistsSourceBal, err := util.FileExists(filepath.Join(balModule, filepath.Base(balFilePath)))
+		if !isExistsSourceBal {
+			fileCopyError := util.CopyFile(balFilePath, filepath.Join(balModule, filepath.Base(balFilePath)))
+			if fileCopyError != nil {
+				return fmt.Errorf("error occurred while copying source bal file, %s", err)
 			}
 		}
 
-		stdoutReader, _ := cmd.StdoutPipe()
-		stdoutScanner := bufio.NewScanner(stdoutReader)
-		go func() {
-			for stdoutScanner.Scan() {
-				fmt.Printf("\r\x1b[2K\033[36m%s\033[m\n", stdoutScanner.Text())
+		if exePath == "" {
+			mntImageDir := strings.Replace(imageDir, util.UserHomeDir(), "/home/cellery", 1)
+			err = CreateBallerinaConf(string(iName), verboseMode, mntImageDir, string(dependencyLinksJson), envVars, testsRoot)
+			if err != nil {
+				return fmt.Errorf("error occurred while creating the ballerina.conf, %s", err)
 			}
-		}()
-		stderrReader, _ := cmd.StderrPipe()
-		stderrScanner := bufio.NewScanner(stderrReader)
-		go func() {
-			for stderrScanner.Scan() {
-				fmt.Printf("\r\x1b[2K\033[36m%s\033[m\n", stderrScanner.Text())
+			err = RunDockerTelepresenceTests(incell, cmd, testsRoot, imageDir, instanceName)
```

Comment:
Do we need these functions to be public?

Reply:
These should be improved to use the ballerina executor. Will be fixed in a separate PR.